Added Tutorial Writeup-LCA

RohitTaparia opened the pull request:
## Change Summary

Issue #241 
Added the tutorial write up for LCA, with walk-through of two LC problems. Please review this PR.


## Checklist

**If you haven't fulfilled the below requirements or even delete the entire checklist, your PR won't be reviewed and will be closed without notice.**

### General 

- [x] This Pull Request is all my own work. (You'll be blacklisted if you are caught for plagiarism.)
- [x] I've read [CONTRIBUTING.md](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md)
- [x] I've applied LaTex for all variables, formulas and time / space complexity instead of using backticks
- [x] I've started the app locally and verified all the content and all links (if applicable) are accessible correctly
- [x] I've included Complexity Analysis (Time Complexity & Space Complexity).
- [x] I've written my explanation well and it is easy to understand for beginners

### Tutorial

- [x] I've read and followed the [Tutorial Template](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md#tutorial-template)
- [x] I've explained my topic well with 2 - 3 LC problems and no external problems are used. 
- [x] I've provided the full working solutions to the problems used in this tutorial.
- [x] I've provided suggested problems at the end with the given format. See [here](https://raw.githubusercontent.com/wingkwong/leetcode-the-hard-way/main/tutorials/math/number-theory/binary-exponentiation.md) as an example. If the target solution is not available, leave `solutionLink` blank.
- [x] I've given credits / references if I use external resources. (For an image, give credit under it. Otherwise, add a new section called References at the end (after Suggested Problems).)

wingkwong reviewed (changes requested):
I just skim through once. Here's some comments.

1. solutionLink is missing in suggestedProblems. if it doesn't have solution, leave the content blank.
2. please check your comment indentation. 
3. comments should start with lower case and a space after `//`
4. in your c++ implementation, use the below format as well so that you don't need to mention it is c++ implementation. besides other people can easily add other lang for this part.

````
<Tabs>
<TabItem value="cpp" label="C++">
<SolutionAuthor name="@RohitTaparia"/>
```cpp
// ...
```
</TabItem>
</Tabs>
````
5. I haven't read deeply but somehow your code in c++ implementation section doesn't match with the solutions in the examples.

RohitTaparia commented:
Thanks for the review @wingkwong  .
The solution does not match the implementation when seen from above, because I have used an iterative stack-based implementation in that section, whereas I have used recursion in the solutions hence I had no need to use stacks. The way of implementation is different but the logic still remains the same. So should I change it?

I have corrected the comments as you said., and changed the format of C++ implementation.

For the solution link, as the solutions are missing, do you mean to say that I should create the solution files and link them and leave the solution content blank in them?

wingkwong commented:
1. I think you can also include recursive implementation as well so that it won't confuse the readers.
2. I don't see you made any changes after my comment
3. Use the below one. If `solutionLink` is "", then it will render `N/A` automatically.

```
export const suggestedProblems = [
  {
    "problemName": "1123. Lowest Common Ancestor of Deepest Leaves",
    "difficulty": "Medium",
    "leetCodeLink": "https://leetcode.com/problems/lowest-common-ancestor-of-deepest-leaves/",
    "solutionLink": ""
  },
  {
    "problemName": "235. Lowest Common Ancestor of a Binary Search Tree",
    "difficulty": "Medium",
    "leetCodeLink": "https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/",
    "solutionLink": ""
  }
]
```

RohitTaparia commented:
@wingkwong I have made the required changes, can you please review it once?

RohitTaparia commented:
Hello @wingkwong , can you please merge or review and suggest me other changes if wanted. Thanks

wingkwong commented:
Sorry. Still busy these days. Will take a look when I'm free. Thanks.

RohitTaparia commented:
Hello @wingkwong , if you have time can you merge or suggest me changes to this PR. Actually, this was my last PR for Hacktoberfest also, and I have really taken the time and worked on it. As today's the last day for getting PRs merged for Hacktoberfest, I would be grateful if you do so, or suggest me changes which I can make today itself.

wingkwong reviewed (changes requested):
I haven't checked the content deeply yet. Just skim through a bit. 

Here's some comments.

1. The code is not following [K&R Coding style](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr). 
2. After a full stop, please add a space. 
3. There's no explanation on your iterative approach. Beginners won't be able to follow it based on the code.
4. Please add more details in your code as much as possible

You may DM me in Discord. Let's see if we can merge it by today.

RohitTaparia commented:
Hey, @wingkwong . I have made some changes. Also, I have kept the indent spacing as 4 spaces, should I reduce it? Can you please review it again.

wingkwong commented on tutorials/graph-theory/LeastCommonAncestor.md:
````diff
@@ -82,16 +82,14 @@ The same logic can be implemented using recursion, so that we do not need to use
 
 ```cpp
-bool func(Node* root, vector<int>& current_path, int value);
+bool func (Node* root, vector<int>& current_path, int value);
````
why do we need this line?

wingkwong commented on tutorials/graph-theory/LeastCommonAncestor.md:
```diff
@@ -82,16 +82,14 @@ The same logic can be implemented using recursion, so that we do not need to use
 // will return LCA only if both node x, y are present, else -1
-int findLCA(Node* root, int x, int y)
-{
+int findLCA (Node* root, int x, int y){
```
Add the space before the bracket. Please check all.

```cpp
int findLCA (Node* root, int x, int y){
```

```cpp
int findLCA (Node* root, int x, int y) {
```

wingkwong commented on tutorials/graph-theory/LeastCommonAncestor.md:
```diff
@@ -101,8 +99,7 @@ int findLCA(Node* root, int x, int y)
 
-bool func(Node* root, vector<int>& current_path, int value);
-{
+bool func(Node* root, vector<int>& current_path, int value){
```
can we have a better name rather than `func`

wingkwong commented on tutorials/graph-theory/LeastCommonAncestor.md:
```diff
@@ -225,41 +222,48 @@ public:
             return true;  
         path.pop_back(); 
         
 		// trying to find node for right
```
comment indentation is not correct.

RohitTaparia commented:
Heyy @wingkwong , I have made the said changes. Can you please review it again? Thanks.

RohitTaparia commented:
@wingkwong made the changes
